src/main/java/uk/gov/companieshouse/api/accounts/service/impl/ApprovalService.java

```diff
@@ -0,0 +1,125 @@
+        this.approvalRepository = approvalRepository;
+        this.approvalTransformer = approvalTransformer;
+        this.smallFullService = smallFullService;
+        this.keyIdGenerator = keyIdGenerator;
+    }
+
+    @Override
+    public ResponseObject<Approval> create(Approval rest, Transaction transaction,
+        String companyAccountId, String requestId) throws DataException {
+        String selfLink = createSelfLink(transaction, companyAccountId);
+        initLinks(rest, selfLink);
+        rest.setEtag(GenerateEtagUtil.generateEtag());
+        rest.setKind(Kind.APPROVAL.getValue());
+        ApprovalEntity approvalEntity = approvalTransformer.transform(rest);
+
+        final Map<String, Object> debugMap = new HashMap<>();
+        debugMap.put("transaction_id", transaction.getId());
+        debugMap.put("company_accounts_id", companyAccountId);
+
+        String id = generateID(companyAccountId);
+        approvalEntity.setId(id);
+        debugMap.put("id", id);
+
+        try {
+            approvalRepository.insert(approvalEntity);
+        } catch (DuplicateKeyException dke) {
+            LOGGER.errorContext(requestId, dke, debugMap);
+            return new ResponseObject<>(ResponseStatus.DUPLICATE_KEY_ERROR, null);
+        } catch (MongoException me) {
+            DataException dataException = new DataException(
+                "Failed to insert " + ResourceName.SMALL_FULL.getName(), me);
+            LOGGER.errorContext(requestId, dataException, debugMap);
+            throw dataException;
+        }
+
+        smallFullService
+            .addLink(companyAccountId, SmallFullLinkType.APPROVAL, selfLink, requestId);
+
+        return new ResponseObject<>(ResponseStatus.CREATED, rest);
+    }
+
+    @Override
+    public ResponseObject<Approval> findById(String id, String requestId) throws DataException {
+        ApprovalEntity approvalEntity;
+        try {
+            approvalEntity = approvalRepository.findById(id).orElse(null);
+        } catch (MongoException me) {
+            final Map<String, Object> debugMap = new HashMap<>();
+            debugMap.put("id", id);
+            DataException dataException = new DataException("Failed to find Approval", me);
+            LOGGER.errorContext(requestId, dataException, debugMap);
+            throw dataException;
+        }
+
+        if (approvalEntity == null) {
+            return new ResponseObject<>(ResponseStatus.NOT_FOUND);
+        }
+        Approval approval = approvalTransformer.transform(approvalEntity);
+        return new ResponseObject<>(ResponseStatus.FOUND, approval);
+    }
+
+    @Override
+    public String generateID(String companyAccountId) {
+        return keyIdGenerator.generate(companyAccountId + "-" + ResourceName.APPROVAL.getName());
+    }
+
+    public String createSelfLink(Transaction transaction, String companyAccountId) {
+        return transaction.getLinks().get(TransactionLinkType.SELF.getLink()) + "/"
```

Comment:
probably best to use a stringbuilder here?

Reply:
It's easier to read as a String concatenation and is compiled to the same code regardless.

Reply:
Hmm I would have to disagree with the ease of reading as the reason I commented was because it was difficult to read what is being created. If not a `StringBuilder` then a `String.format`. But if you insist its easier to read then that's fair enough.

Reply:
IntelliJ disagrees I started to change the code to StringBuilder and it suggested I change it back to String to make it easier to read.